### PR TITLE
Changes from IP2

### PR DESCRIPTION
**Pull request by fdamken**
Famura will most likely reject this, but it's worth a try?

### CODE REVIEW

File: Pyrado/pyrado/algorithms/meta/idr.py [hunk 1]
```diff
@@ -0,0 +1,159 @@
+# Copyright (c) 2020, Fabio Muratore, Honda Research Institute Europe GmbH, and
```
Comment (famura) [inline Pyrado/pyrado/algorithms/meta/idr.py:1]: Please leave IDR out for now, as I still consider it an ablation of SPDR. We can discuss about this.

File: Pyrado/pyrado/algorithms/meta/sprl.py [hunk 2]
```diff
@@ -108,153 +140,45 @@ def mean(self, mean: to.Tensor):
...
-                pointer += self.dim
-            if self._cov_is_parameter:
-                cov_indices = list(range(pointer, pointer + self.dim))
-                pointer += self.dim
-            return stacked, mean_indices, cov_indices
-
-        return stacked
-
-
 class SPRL(Algorithm):
```
Comment (famura) [inline Pyrado/pyrado/algorithms/meta/sprl.py:182]: Let's rename SPRL to SPDR (affects this file and the training files...)

File: Pyrado/pyrado/domain_randomization/domain_parameter.py [hunk 3]
```diff
@@ -345,7 +358,28 @@ def __init__(
...
+
+    @staticmethod
+    def make_broadening(
+        name: List[str],
+        mean: List[float],
+        init_cov_portion: float = 0.001,
+        target_cov_portion: float = 0.1,
+        clip_lo: float = -pyrado.inf,
+        clip_up: float = pyrado.inf,
+    ) -> "SelfPacedDomainParam":
```
Comment (famura) [inline Pyrado/pyrado/domain_randomization/domain_parameter.py:371]: Add doc string

File: Pyrado/pyrado/domain_randomization/domain_parameter.py [hunk 4]
```diff
@@ -309,34 +315,41 @@ def sample(self, num_samples: int = 1) -> List[to.Tensor]:
...
-        target_cov_chol_flat: to.Tensor,
+        target_cov_flat: to.Tensor,
         init_mean: to.Tensor,
-        init_cov_chol_flat: to.Tensor,
+        init_cov_flat: to.Tensor,
+        clip_lo: float,
+        clip_up: float,
     ):
         """
         Constructor.
```
Comment (famura) [inline Pyrado/pyrado/domain_randomization/domain_parameter.py:327]: no dot (I think)

File: Pyrado/pyrado/environments/pysim/quanser_cartpole.py [hunk 5]
```diff
@@ -341,10 +341,5 @@ def _create_task(self, task_args: dict) -> Task:
         state_des = task_args.get("state_des", np.array([0.0, np.pi, 0.0, 0.0]))
         Q = task_args.get("Q", np.diag([3e-1, 5e-1, 5e-3, 1e-3]))
         R = task_args.get("R", np.diag([1e-3]))
-        rew_fcn = QuadrErrRewFcn(Q, R)
```
Comment (famura) [inline Pyrado/pyrado/environments/pysim/quanser_cartpole.py:344]: So you changed the reward function type, but didn't need to change the values of Q and R?

File: Pyrado/pyrado/utils/bijective_transformation.py [hunk 6]
```diff
@@ -0,0 +1,135 @@
...
+        """
+        raise NotImplementedError()
+
+    @abstractmethod
+    def ensures_non_negativity(self) -> bool:
+        """
+        Whether this transformations ensures non-negativity, i.e., whether :py:meth:`.inverse` only returns non-negative
+        values.
+        """
+        raise NotImplementedError()
```
Comment (famura) [inline Pyrado/pyrado/utils/bijective_transformation.py:72]: Is this really necessary? I mean ppl should know this, right?

File: Pyrado/scripts/training/cth_udr_ppo.py [hunk 7]
```diff
@@ -0,0 +1,118 @@
...
+# OR TECHNICAL UNIVERSITY OF DARMSTADT BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+# SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO,
+# PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS;
+# OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER
+# IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+# ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+# POSSIBILITY OF SUCH DAMAGE.
+
+"""
+Train an agent to solve the HalfCheetah swing-up task using Uniform Domain Randomization.
```
Comment (famura) [inline Pyrado/scripts/training/cth_udr_ppo.py:30]: please look at the other half-cheetah tasks, and copy the description

File: Pyrado/scripts/training/qcp-su_sprl-multidim-all-big-ppo.py [hunk 8]
```diff
@@ -0,0 +1,151 @@
+# Copyright (c) 2020, Fabio Muratore, Honda Research Institute Europe GmbH, and
```
Comment (famura) [inline Pyrado/scripts/training/qcp-su_sprl-multidim-all-big-ppo.py:1]: Change file name to end with ...spdr-ppo.py (no info in there)

We do not want to have two ...spdr-ppo.py files!

File: Pyrado/scripts/training/qcp-su_sprl-multidim-all-big-ppo.py [hunk 9]
```diff
@@ -0,0 +1,151 @@
...
+# OR TECHNICAL UNIVERSITY OF DARMSTADT BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+# SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO,
+# PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS;
+# OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER
+# IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+# ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+# POSSIBILITY OF SUCH DAMAGE.
+
+"""
+Train an agent to solve the Quanser CartPole swing-up task using Proximal Policy Optimization and Self-Paced RL.
```
Comment (famura) [inline Pyrado/scripts/training/qcp-su_sprl-multidim-all-big-ppo.py:30]: Add SPDR here

File: Pyrado/scripts/training/qcp-su_sprl-ppo.py [hunk 10]
```diff
@@ -0,0 +1,150 @@
+# Copyright (c) 2020, Fabio Muratore, Honda Research Institute Europe GmbH, and
```
Comment (famura) [inline Pyrado/scripts/training/qcp-su_sprl-ppo.py:1]: fuse with the other SPDR PPO file

File: Pyrado/scripts/training/qcp-su_udr_ppo.py [hunk 11]
```diff
@@ -0,0 +1,128 @@
+# Copyright (c) 2020, Fabio Muratore, Honda Research Institute Europe GmbH, and
```
Comment (famura) [inline Pyrado/scripts/training/qcp-su_udr_ppo.py:1]: Rename file to ...udr-ppo.py (not _)

File: Pyrado/scripts/training/qcp-su_udr_ppo.py [hunk 12]
```diff
@@ -0,0 +1,128 @@
...
+# OR TECHNICAL UNIVERSITY OF DARMSTADT BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+# SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO,
+# PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS;
+# OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER
+# IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+# ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+# POSSIBILITY OF SUCH DAMAGE.
+
+"""
+Train an agent to solve the CartPole swing-up task using Uniform Domain Randomization.
```
Comment (famura) [inline Pyrado/scripts/training/qcp-su_udr_ppo.py:30]: Please add SPDR

File: Pyrado/scripts/training/qq-su_idr-multidim-big-ppo.py [hunk 13]
```diff
@@ -0,0 +1,143 @@
+# Copyright (c) 2020, Fabio Muratore, Honda Research Institute Europe GmbH, and
```
Comment (famura) [inline Pyrado/scripts/training/qq-su_idr-multidim-big-ppo.py:1]: please leave IDR out for now

File: Pyrado/scripts/training/qq-su_idr-ppo.py [hunk 14]
```diff
@@ -0,0 +1,135 @@
+# Copyright (c) 2020, Fabio Muratore, Honda Research Institute Europe GmbH, and
```
Comment (famura) [inline Pyrado/scripts/training/qq-su_idr-ppo.py:1]: please leave IDR out for now

File: Pyrado/scripts/training/qq-su_npdr_gaussian-prior.py [hunk 15]
```diff
@@ -111,7 +112,7 @@
     }
 
     # Transform the domain parameter space
-    env_sim = SqrtDomainParamTransform(env_sim, [dp_name for dp_name in dp_mapping.values()])
+    env_sim = DomainParamTransform(env_sim, [dp_name for dp_name in dp_mapping.values()], SqrtTransformation())
```
Comment (famura) [inline Pyrado/scripts/training/qq-su_npdr_gaussian-prior.py:115]: Please check if this script runs

File: Pyrado/scripts/training/qq-su_sprl-multdim-all-ppo.py [hunk 16]
```diff
@@ -0,0 +1,151 @@
+# Copyright (c) 2020, Fabio Muratore, Honda Research Institute Europe GmbH, and
```
Comment (famura) [inline Pyrado/scripts/training/qq-su_sprl-multdim-all-ppo.py:1]: Let's have only one qq-su_spdr-ppo.py file. Maybe work with args from the command line to run different flavors.

File: Pyrado/scripts/training/qq-su_sprl-multdim-big-ppo.py [hunk 17]
```diff
@@ -0,0 +1,147 @@
+# Copyright (c) 2020, Fabio Muratore, Honda Research Institute Europe GmbH, and
```
Comment (famura) [inline Pyrado/scripts/training/qq-su_sprl-multdim-big-ppo.py:1]: Again, only one qq-su_spdr-ppo.py file. Maybe work with args from the command line to run different flavors.

File: Pyrado/scripts/training/qq-su_sprl-multdim-logtrans-ppo.py [hunk 18]
```diff
@@ -0,0 +1,158 @@
+# Copyright (c) 2020, Fabio Muratore, Honda Research Institute Europe GmbH, and
```
Comment (famura) [inline Pyrado/scripts/training/qq-su_sprl-multdim-logtrans-ppo.py:1]: Again, only one qq-su_spdr-ppo.py file. Maybe work with args from the command line to run different flavors.

File: Pyrado/scripts/training/qq-su_sprl-multdim-masses-lengths-ppo.py [hunk 19]
```diff
@@ -0,0 +1,149 @@
+# Copyright (c) 2020, Fabio Muratore, Honda Research Institute Europe GmbH, and
```
Comment (famura) [inline Pyrado/scripts/training/qq-su_sprl-multdim-masses-lengths-ppo.py:1]: Again, only one qq-su_spdr-ppo.py file. Maybe work with args from the command line to run different flavors.

File: Pyrado/scripts/training/qq-su_sprl-multdim-ppo.py [hunk 20]
```diff
@@ -0,0 +1,168 @@
+# Copyright (c) 2020, Fabio Muratore, Honda Research Institute Europe GmbH, and
```
Comment (famura) [inline Pyrado/scripts/training/qq-su_sprl-multdim-ppo.py:1]: Again, only one qq-su_spdr-ppo.py file. Maybe work with args from the command line to run different flavors.

File: Pyrado/scripts/training/qq-su_sprl-ppo.py [hunk 21]
```diff
@@ -45,6 +45,7 @@
 from pyrado.policies.feed_back.fnn import FNNPolicy
```
Comment (famura) [inline Pyrado/scripts/training/qq-su_sprl-ppo.py:45]: Again, only one qq-su_spdr-ppo.py file. Maybe work with args from the command line to run different flavors.

File: Pyrado/scripts/training/qq-su_udr-big_ppo.py [hunk 22]
```diff
@@ -0,0 +1,137 @@
+# Copyright (c) 2020, Fabio Muratore, Honda Research Institute Europe GmbH, and
```
Comment (famura) [inline Pyrado/scripts/training/qq-su_udr-big_ppo.py:1]: Let's have only one qq-su_udr-ppo.py file. Maybe work with args from the command line to run different flavors.

File: Pyrado/scripts/training/qq-su_udr-grav_ppo.py [hunk 23]
```diff
@@ -0,0 +1,132 @@
+# Copyright (c) 2020, Fabio Muratore, Honda Research Institute Europe GmbH, and
```
Comment (famura) [inline Pyrado/scripts/training/qq-su_udr-grav_ppo.py:1]: Again, only one qq-su_udr-ppo.py file. Maybe work with args from the command line to run different flavors.

File: Pyrado/scripts/training/qq-su_udr_ppo.py [hunk 24]
```diff
@@ -49,16 +49,21 @@
```
Comment (famura) [inline Pyrado/scripts/training/qq-su_udr_ppo.py:49]: Again, only one qq-su_udr-ppo.py file. Maybe work with args from the command line to run different flavors.

File: Pyrado/tests/environment_wrappers/test_combination.py [hunk 25]
```diff
@@ -58,7 +58,7 @@ def test_combination_wrappers_domain_params(env: SimEnv):
     env_do = GaussianObsNoiseWrapper(
         env_d, noise_std=2 * np.ones(env_d.obs_space.shape), noise_mean=3 * np.ones(env_d.obs_space.shape)
     )
-    env_dot = LogDomainParamTransform(env_do, mask=list(env_do.supported_domain_param))
+    env_dot = DomainParamTransform(env_do, mask=list(env_do.supported_domain_param), transformation=LogTransformation())
```
Comment (famura) [inline Pyrado/tests/environment_wrappers/test_combination.py:61]: Did you run this test manually?

**Review by famura — CHANGES_REQUESTED** [review]
Nice additions & reworks. I think we should keep the number of training scripts as low as possible. I also annotated some questions.

Thank you for your contribution :+1:

File: Pyrado/pyrado/algorithms/meta/sprl.py (hunk 2 above)
Comment (fdamken) [reply Pyrado/pyrado/algorithms/meta/sprl.py:182]: I'd not do this as SPDR only arises from SPRL depending on the target distribution. This class implements vanilla SPRL.

File: Pyrado/pyrado/utils/bijective_transformation.py (hunk 6 above)
Comment (fdamken) [reply Pyrado/pyrado/utils/bijective_transformation.py:72]: This was used by the sprl implementation to decide whether a non-negativity constraint has to be applied or not. I'd leave it in 'cause it doesn't hurt.

File: Pyrado/tests/environment_wrappers/test_combination.py (hunk 25 above)
Comment (fdamken) [reply Pyrado/tests/environment_wrappers/test_combination.py:61]: yes

File: Pyrado/scripts/training/qq-su_npdr_gaussian-prior.py (hunk 15 above)
Comment (fdamken) [reply Pyrado/scripts/training/qq-su_npdr_gaussian-prior.py:115]: It does not, but neither does it on master.

```
Traceback (most recent call last):
  File "SimuRLacra/Pyrado/scripts/training/qq-su_npdr_gaussian-prior.py", line 204, in <module>
    **algo_hparam,
  File "SimuRLacra/Pyrado/pyrado/algorithms/meta/npdr.py", line 69, in __init__
    super().__init__(*args, num_checkpoints=3, init_checkpoint=-1, **kwargs)
  File "SimuRLacra/Pyrado/pyrado/algorithms/meta/sbi_base.py", line 240, in __init__
    self._setup_sbi(prior=prior)
  File "SimuRLacra/Pyrado/pyrado/algorithms/meta/sbi_base.py", line 303, in _setup_sbi
    self._sbi_simulator, self._sbi_prior = prepare_for_sbi(rollout_sampler, prior)
  File ".anaconda3/envs/pyrado/lib/python3.7/site-packages/sbi/utils/user_input_checks.py", line 501, in prepare_for_sbi
    check_sbi_inputs(simulator, prior)
  File ".anaconda3/envs/pyrado/lib/python3.7/site-packages/sbi/utils/user_input_checks.py", line 518, in check_sbi_inputs
    simulation = simulator(theta)
  File ".anaconda3/envs/pyrado/lib/python3.7/site-packages/sbi/utils/user_input_checks.py", line 434, in batch_loop_simulator
    xs = list(map(simulator, theta))
  File ".anaconda3/envs/pyrado/lib/python3.7/site-packages/sbi/utils/user_input_checks.py", line 399, in pytorch_simulator
    return torch.as_tensor(simulator(theta), dtype=float32)
  File "SimuRLacra/Pyrado/pyrado/sampling/sbi_rollout_sampler.py", line 317, in __call__
    check_domain_params(ro_sim, dp_value, self.dp_names)
  File "SimuRLacra/Pyrado/pyrado/sampling/sbi_rollout_sampler.py", line 545, in check_domain_params
    msg="The domain parameters after the rollouts are not identical to the ones commanded by the sbi!"
pyrado.utils.exceptions.ValueErr: The domain parameters after the rollouts are not identical to the ones commanded by the sbi!

Process finished with exit code 1

```

File: Pyrado/pyrado/algorithms/meta/idr.py (hunk 1 above)
Comment (fdamken) [reply Pyrado/pyrado/algorithms/meta/idr.py:1]: I added a comment on this fact but I'd leave it as a baseline for future research.

**Review by famura — APPROVED** [review]

**Pull request merged**